pkgs/applications/video/handbrake/default.nix

```diff
@@ -14,20 +14,31 @@
   libiconv, fribidi, fontconfig, freetype, libass, jansson, libxml2, harfbuzz,
   # Optical media
   libdvdread, libdvdnav, libdvdcss, libbluray,
-  useGtk ? true, wrapGAppsHook ? null,
-                 intltool ? null,
-                 glib ? null,
-                 gtk3 ? null,
-                 libappindicator-gtk3 ? null,
-                 libnotify ? null,
-                 gst_all_1 ? null,
-                 dbus-glib ? null,
-                 udev ? null,
-                 libgudev ? null,
-                 hicolor-icon-theme ? null,
+  # Darwin-specific
+  AudioToolbox ? null,
+  Foundation ? null,
+  libobjc ? null,
+  VideoToolbox ? null,
+  # GTK. This is not available on Darwin, since libappindicator-gtk3 is not
+  # supported there.
+  useGtk ? !stdenv.isDarwin, wrapGAppsHook ? null,
```

Comment:
Is your explanation in Pull Request message is quite, sends in a different direction, that it is in the code.

Can you explain adding ```!stdenv.isDarwin``` here.

If I understand correctly, code now says if system is Darwin - then ```!stdenv.isDarwin``` equals ```false```, which disables derivation dependencies:
```nix
{
buildInputs = []
  ++ lib.optionals useGtk [
    glib gtk3 libappindicator-gtk3 libnotify
    gst_all_1.gstreamer gst_all_1.gst-plugins-base dbus-glib udev
    libgudev hicolor-icon-theme
  ];
}
```

And disables GTK:
```nix
(if useGtk          then "--disable-gtk-update-checks" else "--disable-gtk")
```
during build.

Do you want to make Handbrake on macOS headless (in favour of this shows
```nix
(if stdenv.isDarwin then "--disable-xcode"             else ""
```
, which disables GUI build). Or it _somehow_ uses a different GUI toolkit?

You stated that this "Fix handbrake build by adding a dependency on lzma (for all platforms) and adding some Darwin-specific libraries and patches.", it seems quite misleading.

Did you tried to work-around lack of `libappindicator-gtk3` and what were the results?

AFAIR `libappindicator-gtk3` is a tiny functionality and an optional dependency that I kept in closure, I only work in the Linux space. Are you positive that because of GTK build on macOS is not possible and we required to throw away baby with bathwater (of course I also refer to [official build documentation](https://handbrake.fr/docs/en/1.2.0/developer/build-mac.html)?

Sorry for such a long list of remarks and questions, try to provide closer description to changes.

Also:
```nix
# GTK. This is not available on Darwin, since libappindicator-gtk3 is not supported there.
```
Does it mean that this code block is not available, or GTK is not available?

To sum-up:
1. Is `libappindicator-gtk3` is really major required dependency of build on macOS, then we required to make only headless build?
2. Maybe I missed or do not know some information - then please provide more information.

Reply:
Maybe I did throw the baby out with the bathwater 🙂 Your concerns are totally fair and I apologize for not including more detail in the pull request.

I did try to make the libappindicator-gtk3 dependency optional, but then I faced the problem that Handbrake still had a transitive dependency on systemd (which is not available on macOS). I’m not sure which of Handbrake’s dependencies introduced this transitive dependency but I did find that completely disabling GTK support on macOS took care of the problem. (It’s not clear to me whether Handbrake supports being built without libappindicator-gtk3.)

I *also* could not figure out how to build the native macOS GUI. Removing the `--disable-xcode` flag let the package build but the only file in the resulting derivation was still `HandbrakeCLI`. (In retrospect, I’m not sure what prompted me to add this flag in the first place—the package now seems to build equally well regardless of whether or not it’s specified. But the native GUI is not present in either case.)

My assumption was that most Nix users on macOS would prefer a headless Handbrake, at least compared to the status quo, which was that the package was completely unavailable on macOS. I decided that the best way forward would be to completely disable the GTK portion of the package when building under macOS.

I can think of the following ways to resolve this:

1. Remove all Handbrake changes from this PR.
2. Update the PR so that the only change to Handbrake is to add lzma and the Darwin libraries as dependencies. Users on macOS will need to set `useGtk = false` in config.nix in order to install Handbrake.
3. Leave the PR as is. Users on macOS will only be able to install headless Handbrake.
4. Hold the PR until it offers the native Handbrake GUI on macOS.
5. Hold the PR until it offers the GTK GUI on macOS. (Is this possible?)

I’ll leave it to you to decide which resolution would be best.

Reply:
@Anton-Latukha Any thoughts on this?

Reply:
> then I faced the problem that Handbrake still had a transitive dependency on systemd (which is not available on macOS). I’m not sure which of Handbrake’s dependencies introduced this transitive dependency 

This derivation needs to be found. Consider using `nix why-depends`.

Reply:
The dependency chain seems to be Handbrake → various parts of GTK+3 → libatk → at-spi2 → dbus → systemd.

<details>
<summary><code>nix why-depends --all nixpkgs.handbrake nixpkgs.systemd</code></summary>

    /nix/store/5kainpmfmb0i12a73i6zn08piqyxvv7m-handbrake-1.2.2
    ╚═══bin/.ghb-wrapped: …IL_56.LIBAVFORMAT_58./nix/store/hg8r2zrmpkx3al9hsvnkrxvr4rifl54k-gtk+3-3.24.8/lib:/nix/store/164…
        => /nix/store/hg8r2zrmpkx3al9hsvnkrxvr4rifl54k-gtk+3-3.24.8pkx3al9hsvnkrxvr4rifl54k-gtk+3-3.24.8/share/gsettings-sc…
        bin/ghb: …xport XDG_DATA_DIRS='/nix/store/hg8r2zrmpkx3al9hsvnkrxvr4rifl54k-gtk+3-3.24.8/share/gsettings-sc…
    
        ╚═══bin/gtk-launch: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-am-et.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-cedilla.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-cyrillic-translit.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-inuktitut.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-ipa.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-multipress.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-thai.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-ti-er.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-ti-et.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-viqr.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-wayland.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-waylandgtk.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/immodules/im-xim.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/printbackends/libprintbackend-cups.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/printbackends/libprintbackend-file.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/gtk-3.0/3.0.0/printbackends/libprintbackend-lpr.so: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            => /nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.08yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
            lib/libgtk-3.so.0.2404.4: …w2w23-atk-2.32.0/lib:/nix/store/hp9m8yl53dmgydini31l71x8g3407gmq-at-spi2-atk-2.32.0/lib:/nix/sto…
    
            ╚═══lib/libatk-bridge-2.0.so.0.0.0: …j8lg-glib-2.60.3/lib:/nix/store/jlnmaraahxbpnygmhjm1qzsyn8n2b9m8-at-spi2-core-2.32.1/lib:/nix/st…
                => /nix/store/jlnmaraahxbpnygmhjm1qzsyn8n2b9m8-at-spi2-core-2.32.1
                ╚═══libexec/at-spi-bus-launcher: …a boolean but got %s./nix/store/k02nmqccdqhyk00k4a6hr20kv1f5v6pq-dbus-1.12.16/bin/dbus-daemon...…
                    => /nix/store/k02nmqccdqhyk00k4a6hr20kv1f5v6pq-dbus-1.12.16
                    ╚═══etc/systemd/user/dbus.socket: …t/bus.ExecStartPost=-/nix/store/hach8csknszq07jfv8bdgqhgwmj7bcz7-systemd-242/bin/systemctl --use…
                        => /nix/store/hach8csknszq07jfv8bdgqhgwmj7bcz7-systemd-242
</details>

Reply:
@bdesham

Ooops, really sorry for the long pause.

Preface: I myself learned quickly that small PRs are way better then complex epic ones.

1. It would be easier for you to merge changes, if you branch-out and separately PR changes to different packages. If your package fixes rely on each-other - you can keep the main develop branch for yourself, and cherry-pick your main develop branch changes into different branches.

Some while back I touched `ATK` and `at-spi-core2`, since they implement accessibility specification and use `dbus` I wonder why they require `systemd`.

2. I am currently researching the question about dependency on `systemd`, feel free to solve also.

Reply:
I had picked the ffmpeg and mp4v2 stuff around a week ago.

Reply:
Situation of dependency on `systemd` is covered in three comments here https://github.com/NixOS/nixpkgs/issues/62421#issuecomment-497988202

Reply:
@bdesham Thanks for sharing story of experience.
Since now the source of problem is known - current changes are reasonable. Would fix GUI when upstream issue for it is resolved.

Reply:
I don't think that libappindicator-gtk3 is needed by handbrake at all! I can't find it in any build script. Perhaps we can just remove that everywhere.

libgudev is an optional dependency, but can be just added on linux. Otherwise gtk support should just work.